src/storage/txn/actions/check_data_constraint.rs

```diff
@@ -0,0 +1,91 @@
+    // 1.The current write type is `PUT`
+    // 2.The current write type is `Rollback` or `Lock`, and the key have an older version.
+    if write.write_type == WriteType::Put || txn.key_exist(&key, write_commit_ts.prev())? {
+        return Err(ErrorInner::AlreadyExist { key: key.to_raw()? }.into());
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::storage::{TestEngineBuilder, Engine};
+    use txn_types::{TimeStamp, Write, WriteType, Key, Value};
+    use concurrency_manager::ConcurrencyManager;
+    use crate::storage::mvcc::{ErrorInner, MvccTxn, Result as MvccResult};
+    use crate::storage::txn::actions::check_data_constraint::check_data_constraint;
+
+    #[test]
+    fn test_check_data_constraint() {
+        let engine = TestEngineBuilder::new().build().unwrap();
+        let cm = ConcurrencyManager::new(42.into());
+        let snapshot = engine.snapshot(Default::default()).unwrap();
+        let mut txn = MvccTxn::new(snapshot, TimeStamp::new(2), true, cm);
+        txn.put_write(Key::from_raw(b"a"), TimeStamp::new(5), Value::from(b"b".to_vec()));
+
+
+        struct Case {
+            expected: MvccResult<()>,
+
+            should_not_exist: bool,
+            write: Write,
+            write_commit_ts: TimeStamp,
+            key: Key,
+        }
+
+        let cases = vec![
+            // todo: explain these cases and add more
+            Case {
+                expected: Ok(()),
+                should_not_exist: false,
+                write: Write::new(WriteType::Put, TimeStamp::new(10), None),
+                write_commit_ts: Default::default(),
+                key: Key::from_raw(b"a"),
+            },
+            Case {
+                expected: Ok(()),
+                should_not_exist: true,
+                write: Write::new(WriteType::Delete, TimeStamp::new(10), None),
+                write_commit_ts: Default::default(),
+                key: Key::from_raw(b"a"),
+            },
+            Case {
+                expected: Err(ErrorInner::AlreadyExist { key: b"a".to_vec() }.into()),
+                should_not_exist: true,
+                write: Write::new(WriteType::Put, TimeStamp::new(10), None),
+                write_commit_ts: Default::default(),
+                key: Key::from_raw(b"a"),
+            },
+            Case {
+                expected: Err(ErrorInner::AlreadyExist { key: b"a".to_vec() }.into()),
+                should_not_exist: true,
+                write: Write::new(WriteType::Put, TimeStamp::new(10), None),
+                write_commit_ts: TimeStamp::new(15),
+                key: Key::from_raw(b"a"),
+            },
+        ];
+
+        for case in cases {
+            let Case { expected, should_not_exist, write, write_commit_ts, key } = case;
+            let result = check_data_constraint(&mut txn, should_not_exist, &write, write_commit_ts, &key);
+            assert_eq!(format!("{:?}", expected), format!("{:?}", result));
```

Comment:
Should all error types derive `Eq` so we can `assert_eq!(expected, result)` directly here?

Reply:
We could try, maybe it is difficult?

Reply:
There are many Error types in the std library which don't impl `Eq`, which made `#[derive(Eq, PartialEq)]` directly impossible.
I'll have a try to `impl Eq` on `mvcc::ErrorInner` we used here manually.